File: 46-permutations/permutations.java
class Solution {
    List<List<Integer>> permutations;
    List<Integer> nums;

    public List<List<Integer>> permute(int[] nums) {
        permutations = new ArrayList<>();
        this.nums = new ArrayList<>();
        for (int i: nums) {
            this.nums.add(i);   // Create arraylist from array
        }
        List<Integer> current = new ArrayList<>();
        permuteHelper(current);
        return permutations;
    }

    public void permuteHelper(List<Integer> curr) {
        if (nums.size() == 0) {
            permutations.add(new ArrayList<>(curr));
            return;
        }

        for (int i = 0; i < nums.size(); i++) {
            int x = nums.remove(i);
            curr.add(x);
            permuteHelper(curr);
            curr.remove(curr.size()-1);
            nums.add(i, x);
        }
    }
}

File: 46-permutations/permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        if nums == []:
            return []

        if len(nums) == 1:  # If list has only one element
            return [nums]

        permuteList = []
        self.permuteHelper(nums, [], permuteList)
        return permuteList

    def permuteHelper(self, nums: List[int], permutation: List[int], permutations: List[List[int]]) -> None:
        if nums == []:
            permutations.append(permutation.copy())
            return
        
        for i in range(len(nums)):
            x = nums.pop(i)
            self.permuteHelper(nums, permutation+[x], permutations)
            nums.insert(i, x)